graveler/committed/tree/writer.go

```diff
@@ -55,90 +61,87 @@ func (w *writer) WriteRecord(record graveler.ValueRecord) error {
 	}
 	err = w.partWriter.WriteRecord(committed.Record{Key: committed.Key(record.Key), Value: v})
 	if err != nil {
-		return err
+		return fmt.Errorf("write record to part: %w", err)
 	}
 	w.lastKey = committed.Key(record.Key)
-	breakPoint, err := w.breakPoint(record.Key)
+	breakpoint, err := w.shouldBreakAtKey(record.Key)
 	if err != nil {
 		return err
 	}
-	if breakPoint {
-		if err = w.closeCurrentPart(true); err != nil {
-			return err
-		}
+	if breakpoint {
+		return w.closeCurrentPart()
 	}
 	return nil
 }
 
-func (w *writer) closeCurrentPart(reachedBrakePoint bool) error {
+func (w *GeneralWriter) closeCurrentPart() error {
 	if w.partWriter == nil {
 		return nil
 	}
-	wr, err := w.partWriter.Close()
-	if err != nil {
-		return err
-	}
-	part := Part{
-		ID:                wr.PartID,
-		MinKey:            wr.First,
-		MaxKey:            wr.Last,
-		EstimatedSize:     wr.Count,
-		ReachedBrakePoint: reachedBrakePoint,
-	}
-	err = w.writePartToTree(part)
-	if err != nil {
+	if err := w.batchWriteCloser.CloseWriterAsync(w.partWriter); err != nil {
 		return err
 	}
 	w.partWriter = nil
 	return nil
 }
 
-func (w *writer) AddPart(part Part) error {
-	if w.lastKey != nil && bytes.Compare(part.MinKey, w.lastKey) < 0 {
-		return ErrUnsorted
+func (w *GeneralWriter) getBatchedParts() ([]Part, error) {
+	wr, err := w.batchWriteCloser.Wait()
+	if err != nil {
+		return nil, fmt.Errorf("batch write closer wait: %w", err)
+	}
+	parts := make([]Part, len(wr))
+	for i, r := range wr {
+		parts[i] = Part{
+			ID:            r.PartID,
+			MinKey:        r.First,
+			MaxKey:        r.Last,
+			EstimatedSize: r.EstimatedSize,
+		}
 	}
-	if err := w.closeCurrentPart(false); err != nil {
-		return err
+	return parts, nil
+}
+
+func (w *GeneralWriter) AddPart(part Part) error {
+	if w.lastKey != nil && bytes.Compare(part.MinKey, w.lastKey) <= 0 {
+		return ErrUnsortedKeys
 	}
-	if err := w.writePartToTree(part); err != nil {
+	if err := w.closeCurrentPart(); err != nil {
 		return err
 	}
 	w.lastKey = part.MaxKey
 	return nil
 }
 
-func (w *writer) SaveTree() (*graveler.TreeID, error) {
-	if err := w.closeCurrentPart(false); err != nil {
+func (w *GeneralWriter) Close() (*graveler.TreeID, error) {
+	if err := w.closeCurrentPart(); err != nil {
 		return nil, err
 	}
-	tree, err := w.treeWriter.Close()
+	parts, err := w.getBatchedParts()
 	if err != nil {
 		return nil, err
 	}
-	treeID := graveler.TreeID(tree.PartID)
-	return &treeID, nil
+	parts = append(parts, w.parts...)
+	sort.Slice(parts, func(i, j int) bool {
+		return bytes.Compare(parts[i].MinKey, parts[j].MinKey) < 0
+	})
```

Comment:
@itaiad200 this is a bit unfortunate, we got the parts in order, sent them over to a batch closer, the batch closer kept information about each of them but lost that order. :-(

Reply:
Also, some parts were passed as is, so sorting here is mandatory.